common/src/test/java/org/conscrypt/MacTest.java

```diff
@@ -112,6 +130,170 @@ public void knownAnswerTest() throws Exception {
+
+                            try {
+                                Mac mac = supplier.get();
+                                mac.update(message);
+                                fail();
+                            } catch (IllegalStateException e) {
+                                // Expected
+                            }
+                            try {
+                                Mac mac = supplier.get();
+                                mac.doFinal(message);
+                                fail();
+                            } catch (IllegalStateException e) {
+                                // Expected
+                            }
+                            try {
+                                Mac mac = supplier.get();
+                                mac.doFinal();
+                                fail();
+                            } catch (IllegalStateException e) {
+                                // Expected
+                            }
+                        }
+                    }
+                });
+
+    }
+
+    @Test
+    public void algorithmParameters() {
+        ServiceTester.test("Mac")
+                .run(new ServiceTester.Test() {
+                    @Override
+                    public void test(final Provider provider, final String algorithm) throws Exception {
+
+                        for (MacSupplier supplier : getMacSuppliers(provider, algorithm)) {
+                            Mac mac = supplier.get();
+                            SecretKeySpec key = findAnyKey(mac.getAlgorithm());
+                            if (key == null) {
+                                // TODO(prb) We should have at least one test vector for every
+                                // MAC in Conscrypt
+                                continue;
+                            }
+
+                            // Equivalent to mac.init(key) - allowed
+                            mac.init(key, null);
+
+                            try {
+                                mac = supplier.get();
+                                mac.init(key, new DummyParameterSpec());
+                                fail();
+                            } catch (InvalidAlgorithmParameterException exception) {
+                                // Expected
+                            }
+                        }
+                    }
+                });
+
+    }
+
+    // Uses of ServiceTester above need to get multiple instances of each Mac to test
+    // failure cases.  This method converts the provided args into a list of suppliers.
+    private List<MacSupplier> getMacSuppliers(final Provider provider, final String algorithm) {
+        List<MacSupplier> suppliers = new ArrayList<>();
+
+        suppliers.add(new MacSupplier() {
+            @Override
+            public Mac get() throws Exception{
+                Mac mac = Mac.getInstance(algorithm);
+                assertEquals(algorithm, mac.getAlgorithm());
```

Comment:
nit: is it possible to extract these checks into separate test case?

Reply:
Yeah, good point, thanks! That structure was leftover from when I had a single `ServiceTester` test calling out to multiple sub-test methods.  I've added a new `ServiceTester` test which just validates creation (and, bonus, initialisation) so the other tests can just use and trust `Mac.getInstance()`.